File: event-sourcing/typescript/src/Account.ts
import type { AccountStatus } from './AccountStatus.js';
import { AccountSummary } from './AccountSummary.js';
import type { AccountEvent, AccountOpened, MoneyDeposited, MoneyWithdrawn } from './Events.js';
import { moneyDeposited, moneyWithdrawn, accountClosed } from './Events.js';
import {
  AccountNotOpenException,
  AccountClosedException,
  InsufficientFundsException,
  InvalidAmountException,
  NonZeroBalanceException,
} from './Exceptions.js';
import { Money } from './Money.js';
import { Transaction } from './Transaction.js';

export class Account {
  private readonly events: AccountEvent[] = [];
  private ownerName = '';
  private _balance: Money = Money.zero;
  private _status: AccountStatus = 'Open';
  private opened = false;

  readonly accountId: string;

  get balance(): Money { return this._balance; }
  get status(): AccountStatus { return this._status; }

  private constructor(accountId: string) {
    this.accountId = accountId;
  }

  static rebuild(events: AccountEvent[]): Account {
    if (events.length === 0) {
      throw new AccountNotOpenException('Account has no events.');
    }

    const first = events[0]!;
    if (first.type !== 'AccountOpened') {
      throw new AccountNotOpenException('Account was never opened.');
    }

    const account = new Account(first.accountId);
    for (const e of events) {
      account.apply(e);
    }
    return account;
  }

  deposit(amount: Money, timestamp: Date): AccountEvent {
    this.ensureOpen();
    if (!amount.isPositive) {
      throw new InvalidAmountException('Deposit amount must be positive.');
    }

    const e = moneyDeposited(this.accountId, amount, timestamp);
    this.apply(e);
    return e;
  }

  withdraw(amount: Money, timestamp: Date): AccountEvent {
    this.ensureOpen();
    if (!amount.isPositive) {
      throw new InvalidAmountException('Withdrawal amount must be positive.');
    }
    if (amount.greaterThan(this._balance)) {
      throw new InsufficientFundsException('Withdrawal amount exceeds current balance.');
    }

    const e = moneyWithdrawn(this.accountId, amount, timestamp);
    this.apply(e);
    return e;
  }

  close(timestamp: Date): AccountEvent {
    this.ensureOpen();
    if (!this._balance.equals(Money.zero)) {
      throw new NonZeroBalanceException('Cannot close account with non-zero balance.');
    }

    const e = accountClosed(this.accountId, timestamp);
    this.apply(e);
    return e;
  }

  transactionHistory(): Transaction[] {
    const transactions: Transaction[] = [];
    let running = Money.zero;
    for (const e of this.events) {
      if (e.type === 'MoneyDeposited') {
        const d = e as MoneyDeposited;
        running = running.plus(d.amount);
        transactions.push(new Transaction(d.timestamp, d.amount, running));
      } else if (e.type === 'MoneyWithdrawn') {
        const w = e as MoneyWithdrawn;
        running = running.minus(w.amount);
        transactions.push(new Transaction(w.timestamp, w.amount.negate(), running));
      }
    }
    return transactions;
  }

  summary(): AccountSummary {
    const txCount = this.events.filter(
      e => e.type === 'MoneyDeposited' || e.type === 'MoneyWithdrawn',
    ).length;
    return new AccountSummary(this.ownerName, this._balance, txCount, this._status);
  }

  balanceAt(pointInTime: Date): Money {
    let balance = Money.zero;
    for (const e of this.events) {
      if (e.timestamp > pointInTime) break;
      if (e.type === 'MoneyDeposited') {
        balance = balance.plus((e as MoneyDeposited).amount);
      } else if (e.type === 'MoneyWithdrawn') {
        balance = balance.minus((e as MoneyWithdrawn).amount);
      }
    }
    return balance;
  }

  transactionsInRange(from: Date, to: Date): Transaction[] {
    const transactions: Transaction[] = [];
    let running = Money.zero;
    for (const e of this.events) {
      if (e.type === 'MoneyDeposited') {
        const d = e as MoneyDeposited;
        running = running.plus(d.amount);
        if (d.timestamp >= from && d.timestamp <= to) {
          transactions.push(new Transaction(d.timestamp, d.amount, running));
        }
      } else if (e.type === 'MoneyWithdrawn') {
        const w = e as MoneyWithdrawn;
        running = running.minus(w.amount);
        if (w.timestamp >= from && w.timestamp <= to) {
          transactions.push(new Transaction(w.timestamp, w.amount.negate(), running));
        }
      }
    }
    return transactions;
  }

  private apply(e: AccountEvent): void {
    switch (e.type) {
      case 'AccountOpened': {
        const opened = e as AccountOpened;
        this.ownerName = opened.ownerName;
        this.opened = true;
        this._status = 'Open';
        break;
      }
      case 'MoneyDeposited':
        this._balance = this._balance.plus((e as MoneyDeposited).amount);
        break;
      case 'MoneyWithdrawn':
        this._balance = this._balance.minus((e as MoneyWithdrawn).amount);
        break;
      case 'AccountClosed':
        this._status = 'Closed';
        break;
    }
    this.events.push(e);
  }

  private ensureOpen(): void {
    if (!this.opened) {
      throw new AccountNotOpenException('Account was never opened.');
    }
    if (this._status === 'Closed') {
      throw new AccountClosedException('Account is closed.');
    }
  }
}

File: event-sourcing/python/src/event_sourcing/account.py
from __future__ import annotations

from datetime import datetime

from .account_status import AccountStatus
from .account_summary import AccountSummary
from .events import AccountEvent, AccountOpened, MoneyDeposited, MoneyWithdrawn, AccountClosed
from .exceptions import (
    AccountNotOpenException,
    AccountClosedException,
    InsufficientFundsException,
    InvalidAmountException,
    NonZeroBalanceException,
)
from .money import Money
from .transaction import Transaction


class Account:
    def __init__(self, account_id: str) -> None:
        self._account_id = account_id
        self._owner_name = ""
        self._balance = Money.zero()
        self._status = AccountStatus.OPEN
        self._opened = False
        self._events: list[AccountEvent] = []

    @property
    def account_id(self) -> str:
        return self._account_id

    @property
    def balance(self) -> Money:
        return self._balance

    @property
    def status(self) -> AccountStatus:
        return self._status

    @staticmethod
    def rebuild(events: list[AccountEvent]) -> Account:
        if not events:
            raise AccountNotOpenException("Account has no events.")

        first = events[0]
        if not isinstance(first, AccountOpened):
            raise AccountNotOpenException("Account was never opened.")

        account = Account(first.account_id)
        for e in events:
            account._apply(e)
        return account

    def deposit(self, amount: Money, timestamp: datetime) -> AccountEvent:
        self._ensure_open()
        if not amount.is_positive:
            raise InvalidAmountException("Deposit amount must be positive.")

        e = MoneyDeposited(
            account_id=self._account_id, timestamp=timestamp, amount=amount
        )
        self._apply(e)
        return e

    def withdraw(self, amount: Money, timestamp: datetime) -> AccountEvent:
        self._ensure_open()
        if not amount.is_positive:
            raise InvalidAmountException("Withdrawal amount must be positive.")
        if amount > self._balance:
            raise InsufficientFundsException(
                "Withdrawal amount exceeds current balance."
            )

        e = MoneyWithdrawn(
            account_id=self._account_id, timestamp=timestamp, amount=amount
        )
        self._apply(e)
        return e

    def close(self, timestamp: datetime) -> AccountEvent:
        self._ensure_open()
        if self._balance != Money.zero():
            raise NonZeroBalanceException(
                "Cannot close account with non-zero balance."
            )

        e = AccountClosed(account_id=self._account_id, timestamp=timestamp)
        self._apply(e)
        return e

    def transaction_history(self) -> list[Transaction]:
        transactions: list[Transaction] = []
        running = Money.zero()
        for e in self._events:
            if isinstance(e, MoneyDeposited):
                running = running + e.amount
                transactions.append(Transaction(e.timestamp, e.amount, running))
            elif isinstance(e, MoneyWithdrawn):
                running = running - e.amount
                transactions.append(Transaction(e.timestamp, -e.amount, running))
        return transactions

    def summary(self) -> AccountSummary:
        tx_count = sum(
            1 for e in self._events if isinstance(e, (MoneyDeposited, MoneyWithdrawn))
        )
        return AccountSummary(self._owner_name, self._balance, tx_count, self._status)

    def balance_at(self, point_in_time: datetime) -> Money:
        balance = Money.zero()
        for e in self._events:
            if e.timestamp > point_in_time:
                break
            if isinstance(e, MoneyDeposited):
                balance = balance + e.amount
            elif isinstance(e, MoneyWithdrawn):
                balance = balance - e.amount
        return balance

    def transactions_in_range(
        self, from_time: datetime, to_time: datetime
    ) -> list[Transaction]:
        transactions: list[Transaction] = []
        running = Money.zero()
        for e in self._events:
            if isinstance(e, MoneyDeposited):
                running = running + e.amount
                if from_time <= e.timestamp <= to_time:
                    transactions.append(Transaction(e.timestamp, e.amount, running))
            elif isinstance(e, MoneyWithdrawn):
                running = running - e.amount
                if from_time <= e.timestamp <= to_time:
                    transactions.append(Transaction(e.timestamp, -e.amount, running))
        return transactions

    def _apply(self, e: AccountEvent) -> None:
        if isinstance(e, AccountOpened):
            self._owner_name = e.owner_name
            self._opened = True
            self._status = AccountStatus.OPEN
        elif isinstance(e, MoneyDeposited):
            self._balance = self._balance + e.amount
        elif isinstance(e, MoneyWithdrawn):
            self._balance = self._balance - e.amount
        elif isinstance(e, AccountClosed):
            self._status = AccountStatus.CLOSED
        self._events.append(e)

    def _ensure_open(self) -> None:
        if not self._opened:
            raise AccountNotOpenException("Account was never opened.")
        if self._status == AccountStatus.CLOSED:
            raise AccountClosedException("Account is closed.")
